File: examples/mailer/go/bridges/mailer/host.go
// bridges/mailer/host.go — Go bridge host for the mailer example.
// rusm build generates _runner.go and go.mod (if absent); TinyGo compiles the whole
// bridges/mailer/ package to wasm/bridge-mailer.wasm.
// Set RESEND_API_KEY in the environment or .env before serving.
package main

import (
	"bytes"
	"encoding/json"
	"net/http"
	"os"
)

var (
	apiKey = os.Getenv("RESEND_API_KEY")
	client = &http.Client{}
)

// Send delivers the email via Resend. The generated dispatcher calls this with the WIT
// `message` record serialised into a json.RawMessage — unmarshal into your struct.
func Send(raw json.RawMessage) bool {
	var msg struct {
		To      string `json:"to"`
		Subject string `json:"subject"`
		Body    string `json:"body"`
	}
	if err := json.Unmarshal(raw, &msg); err != nil {
		return false
	}
	payload, _ := json.Marshal(map[string]string{
		"from":    "noreply@example.com",
		"to":      msg.To,
		"subject": msg.Subject,
		"html":    msg.Body,
	})
	req, _ := http.NewRequest("POST", "https://api.resend.com/emails", bytes.NewReader(payload))
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("Content-Type", "application/json")
	resp, err := client.Do(req)
	if err != nil {
		return false
	}
	resp.Body.Close()
	return resp.StatusCode < 300
}

File: examples/mailer/rust/bridges/mailer/host.rs
//! The mailer bridge's Rust host impl — authored in the platform-bridge convention
//! (`impl <iface>::Host for BridgeHost` + `pub fn add_to_linker`). Calls the Resend API
//! via `reqwest`. Set `RESEND_API_KEY` in the environment or `.env` before serving.
//!
//! `rusm build` generates `src/bindings.rs`, `src/bridges.rs`, `wit/`, and the synthesized
//! bindgen world. This file is the only one the app author writes for the bridge's behaviour.

use crate::bindings::app::mailer::smtp;
use rusm_wasm::wasmtime::component::HasSelf;
use rusm_wasm::{wasmtime, BridgeHost, BridgeLinker};

/// Register this bridge into the component linker (called by the generated `bridges::extend`).
pub fn add_to_linker(linker: &mut BridgeLinker) -> wasmtime::Result<()> {
    smtp::add_to_linker::<_, HasSelf<BridgeHost>>(linker, |host| host)
}

impl smtp::Host for BridgeHost {
    async fn send(&mut self, msg: smtp::Message) -> bool {
        let Ok(api_key) = std::env::var("RESEND_API_KEY") else {
            return false;
        };
        reqwest::Client::new()
            .post("https://api.resend.com/emails")
            .bearer_auth(&api_key)
            .json(&serde_json::json!({
                "from":    "noreply@example.com",
                "to":      msg.to,
                "subject": msg.subject,
                "html":    msg.body,
            }))
            .send()
            .await
            .map(|r| r.status().is_success())
            .unwrap_or(false)
    }
}
